File: sampling/C++/RandomWalk.cpp
#include "RandomWalk.h"
#include <time.h>
#include <stdio.h>
#include <iostream>
using namespace std;

PUNGraph RandomWalk::walk(double percent) {
	srand(time(NULL));
  int node_num = graph->GetNodes();
	TIntV all_ids;
	graph->GetNIdV(all_ids);
	target_node_num = ceil(node_num * percent);
  cout << "Sample Target: " << target_node_num << endl;
  sample_g = TUNGraph::New();
  while (sample_g->GetNodes() < target_node_num) {
    int istart = all_ids[rand()%node_num];
    //cout << "Start from node " << istart << endl;
	cout<<"Sampled graph size(target="<<target_node_num<<"): "<<sample_g->GetNodes()<<endl;
    start_journey(istart, all_ids);
  }
	return sample_g;
}

RandomWalk::RandomWalk(PUNGraph g, int cp, int minc) {
	graph = g;
	check_period = cp;
	min_inc = minc;
}

void RandomWalk::start_journey(int istart, const TIntV &all_node_ids) {
	double jumpback_prob = 0.15;
	int last_check_sz = sample_g->GetNodes();
	TUNGraph::TNodeI currentI = graph->GetNI(istart);
	int srcId = istart;
	if (!sample_g->IsNode(istart))
		sample_g->AddNode(istart);
	while (sample_g->GetNodes() < target_node_num) {
		//cout<< sample_g->GetNodes()<<endl;
		for (int i = 0; i < check_period; i++) {
			double prob = (double)rand() / RAND_MAX;
			if (prob < jumpback_prob) {
				currentI = graph->GetNI(istart);
				srcId = istart;
			}
			else {
				int deg = currentI.GetDeg();
				int ineighbor = currentI.GetNbrNId(rand()%deg);
				if (!sample_g->IsNode(ineighbor)) {
					sample_g->AddNode(ineighbor);
					//cout << sample_g->GetNodes() << endl;
				}
				sample_g->AddEdge(srcId, ineighbor);
				srcId = ineighbor;
				currentI = graph->GetNI(ineighbor);
			}
			if (sample_g->GetNodes() >= target_node_num)
				break;
		}
		if (sample_g->GetNodes() - last_check_sz < min_inc)
			break;
		last_check_sz = sample_g->GetNodes();
	}
}

File: sampling/random_walk.py
import networkx as nx
import os
import sys
import random

def start_journey(sample_g, whole_graph, istart, check_period, 
            min_inc, target_node_num):
  jumpback_prob = 0.15
  last_check_sz = sample_g.number_of_nodes()
  current_node = whole_graph.nodes()[istart]
  while sample_g.number_of_nodes() < target_node_num:
    print sample_g.number_of_nodes(), last_check_sz, min_inc
    for i in range(0, check_period):
      if random.random() < jumpback_prob:
        current_node = whole_graph.nodes()[istart]
      else:
        neighbors = [n for n in nx.all_neighbors(whole_graph, current_node)]
        num_neighbors = len(neighbors)
        if num_neighbors == 0:
          break
        ineighbor = random.randint(0, num_neighbors-1)
        sample_g.add_edge(current_node, neighbors[ineighbor])
        current_node = neighbors[ineighbor]
      if sample_g.number_of_nodes() >= target_node_num:
        break

    if sample_g.number_of_nodes() - last_check_sz < min_inc:
      break 
    else:
      last_check_sz = sample_g.number_of_nodes()

def walk(whole_graph, sample_percent, dlm = '\t'):
  node_num = whole_graph.number_of_nodes()
  print 'Done.', node_num, ' nodes in total'
  target_node_num = node_num * sample_percent
  print 'Target: sample', target_node_num, 'nodes'

  sampled_g = nx.Graph()
  check_period = 500
  min_inc = 500 * 0.3
  while sampled_g.number_of_nodes() < target_node_num:
    istart = random.randint(0, node_num-1)
    print istart
    start_journey(sampled_g, whole_graph, istart, check_period, 
        min_inc, target_node_num)
    print 'sampled size', sampled_g.number_of_nodes()
  
  return sampled_g

def ensure_dir(path): 
  if not os.path.exists(path):
    os.makedirs(path)

if __name__ == '__main__':
  input_graph = sys.argv[1]
  dlm = '\t'
  print 'Reading graph', input_graph 
  whole_graph = nx.read_adjlist(input_graph, delimiter = dlm, create_using=nx.Graph())
  num_sample_iterations = 100;
  sample_percents = range(1, 50, 2)   
  output_dir = 'samples/' 
  for p in sample_percents:
    print p
    percent_dir = output_dir + str(p) + '/'
    ensure_dir(percent_dir)
    for i in range(1, num_sample_iterations+1):
      print i
      output_file = percent_dir + str(i) 
      sample_g = walk(whole_graph, float(p)/100, dlm)
      nx.write_adjlist(sample_g, output_file) 
